### PR TITLE
Add isDeleted() to IMessage (re made after failing to use git)

### PR DESCRIPTION
**Pull request by ArsenArsen**
### Prerequisites
* [x] I have read and understood the [contribution guidelines](CONTRIBUTING.md)
* [x] This pull request follows the code style of the project
* [x] I have tested this feature thoroughly
  * [Proof of testing](http://i.imgur.com/9wl3Ilb.png)

**Issues Fixed:** Before these changes were made, trying to edit a deleted message would result in a null pointer exception. What I've done is added isDeleted() boolean and a check to the edit() function. 

### Changes Proposed in this Pull Request
* Add isDeleted()
* Make a check which throws a DiscordException when trying to edit a deleted message

### CODE REVIEW

File: src/main/java/sx/blah/discord/handle/obj/IMessage.java [hunk 1]
```diff
@@ -221,11 +221,11 @@ void addReaction(String emoji) throws MissingPermissionsException, RateLimitExce
 	/**
 	 * Removes a reaction for a user.
 	 *
-	 * @[param user The user
 	 * @param reaction The reaction to remove from
 	 * @throws MissingPermissionsException
 	 * @throws RateLimitException
 	 * @throws DiscordException
+	 * @[param user The user
```
Comment (chrislo27) [inline src/main/java/sx/blah/discord/handle/obj/IMessage.java]: [

File: src/main/java/sx/blah/discord/handle/impl/obj/Message.java [hunk 2]
```diff
@@ -554,4 +565,13 @@ public int hashCode() {
 	public boolean equals(Object other) {
 		return other != null && this.getClass().isAssignableFrom(other.getClass()) && ((IMessage) other).getID().equals(getID());
 	}
+
+	/**
+	 * Sets the CACHED deleted value
```
Comment (chrislo27) [inline src/main/java/sx/blah/discord/handle/impl/obj/Message.java]: Complete sentences.

File: src/main/java/sx/blah/discord/handle/impl/obj/Message.java [hunk 3]
```diff
@@ -122,6 +122,10 @@
 	 * The pattern for matching channel mentions.
 	 */
 	private static final Pattern CHANNEL_PATTERN = Pattern.compile("<#([0-9]+)>");
+	/**
+	 * Cached value of isDeleted()
+	 */
+	private boolean deleted = false;
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Message.java]: Make volatile or a final AtomicBoolean

File: src/main/java/sx/blah/discord/handle/impl/obj/Message.java [hunk 4]
```diff
@@ -169,7 +173,7 @@ public void setContent(String content) {
 	/**
 	 * Sets the CACHED mentions in this message.
 	 *
-	 * @param mentions The new user mentions.
+	 * @param mentions     The new user mentions.
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Message.java]: Why?

File: src/main/java/sx/blah/discord/handle/impl/obj/Message.java [hunk 5]
```diff
@@ -286,10 +290,12 @@ public void reply(String content) throws MissingPermissionsException, RateLimitE
 	public IMessage edit(String content) throws MissingPermissionsException, RateLimitException, DiscordException {
 		if (!this.getAuthor().equals(client.getOurUser()))
 			throw new MissingPermissionsException("Cannot edit other users' messages!");
+		if(isDeleted())
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Message.java]: Add a space after the if

File: src/main/java/sx/blah/discord/handle/obj/IMessage.java [hunk 6]
```diff
@@ -221,11 +221,11 @@ void addReaction(String emoji) throws MissingPermissionsException, RateLimitExce
 	/**
 	 * Removes a reaction for a user.
 	 *
-	 * @[param user The user
 	 * @param reaction The reaction to remove from
 	 * @throws MissingPermissionsException
 	 * @throws RateLimitException
 	 * @throws DiscordException
+	 * @param user The user
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/obj/IMessage.java]: Move this to the appropriate spot

**Comment by ArsenArsen** [comment]
@austinv11 okay, I am working on that.

File: src/main/java/sx/blah/discord/handle/impl/obj/Channel.java [hunk 7]
```diff
@@ -529,7 +516,7 @@ public void unpin(IMessage message) throws RateLimitException, DiscordException,
 
 	@Override
 	public boolean isDeleted() {
-		return deleted;
+		return getClient().getChannelByID(id) != this;
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Channel.java]: Check guild.getChannelByID() instead, it'll be more efficient.

File: src/main/java/sx/blah/discord/handle/impl/obj/Role.java [hunk 8]
```diff
@@ -263,7 +259,7 @@ public boolean isEveryoneRole() {
 
 	@Override
 	public boolean isDeleted() {
-		return deleted;
+		return getClient().getRoleByID(id) != this;
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Role.java]: Again, use the guild object, it'll be more efficient.

File: src/main/java/sx/blah/discord/handle/impl/obj/PrivateChannel.java [hunk 9]
```diff
@@ -167,4 +167,9 @@ public String toString() {
 	public IPrivateChannel copy() {
 		return new PrivateChannel(client, recipient, id);
 	}
+
+	@Override
+	public boolean isDeleted() {
+		return getClient().getChannelByID(id) != this;
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/PrivateChannel.java]: This should just return false. There is no case where a private channel should be "deleted" in the case of how this api handles private channels.

File: src/main/java/sx/blah/discord/handle/impl/obj/Channel.java [hunk 10]
```diff
@@ -516,7 +516,7 @@ public void unpin(IMessage message) throws RateLimitException, DiscordException,
 
 	@Override
 	public boolean isDeleted() {
-		return getClient().getChannelByID(id) != this;
+		return getGuild().getChannelByID(id) != this;
```
Comment (austinv11) [inline src/main/java/sx/blah/discord/handle/impl/obj/Channel.java]: You should add an override to this for VoiceChannel as VoiceChannels are stored in a separate list than the text channels.

**Pull request merged**